jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/PooledJMSConnectionFactory.java

```diff
@@ -0,0 +1,207 @@
+    }
+
+    @Override
+    public TopicConnection createTopicConnection() throws JMSException {
+        return (TopicConnection) createConnection();
+    }
+
+    @Override
+    public TopicConnection createTopicConnection(String userName, String password) throws JMSException {
+        return (TopicConnection) createConnection(userName, password);
+    }
+
+    @Override
+    public void closeConnection(Connection connection) throws JMSConnectorException {
+        PooledConnectionKey key = connectionKeyMap.get(connection);
+
+        if (key != null) {
+            try {
+                keyedObjectPool.returnObject(key, connection);
+            } catch (Exception e) {
+                throw new JMSConnectorException("Failed to return the connection back to the pool", e);
+            }
+        }
+    }
+
+    // PooledObjectFactory implementations
+
+    /**
+     * This is invoked when a new connection object has to be made.
+     *
+     * @param key The connection key
+     * @return Wrapped connection object
+     * @throws Exception Any exception thrown when creating a JMS connection will be thrown
+     */
+    @Override
+    public PooledObject<Connection> makeObject(PooledConnectionKey key) throws Exception {
+        Connection connection;
+
+        String username = key.getUsername();
+        String password = key.getPassword();
+
+        if (username == null && password == null) {
+            connection = super.createConnection();
+        } else {
+
+            connection = super.createConnection(key.getUsername(), key.getPassword());
+        }
+
+        connectionKeyMap.put(connection, key);
+
+        return new DefaultPooledObject<>(connection);
+    }
+
+    /**
+     * This is called when an connection object is invalidated.
+     *
+     * @param key                    The connection key
+     * @param pooledConnectionObject The invalidated wrapped connection object.
+     * @throws Exception Any exception thrown when closing a JMS connection will be thrown
+     */
+    @Override
+    public void destroyObject(PooledConnectionKey key, PooledObject<Connection> pooledConnectionObject) throws
+            Exception {
+        Connection connection = pooledConnectionObject.getObject();
+
+        connectionKeyMap.remove(connection);
+        connection.close();
+    }
+
+    /**
+     * Validates a connection object and verifies it can be used further.
+     *
+     * @param key                    The connection key
+     * @param pooledConnectionObject The wrapped connection object
+     * @return True if the object is valid
+     */
+    @Override
+    public boolean validateObject(PooledConnectionKey key, PooledObject<Connection> pooledConnectionObject) {
+
+        boolean valid = false;
+
+        try {
+            pooledConnectionObject.getObject().start();
+            valid = true;
+        } catch (JMSException e) {
+            log.warn("Connection with username " + key + " is not valid anymore");
```

Comment:
We cannot be sure of this, hence will change the message. Thanks for pointing this out.